### PR TITLE
ASC-331/ASC-597/Display list of Molecule Orders

### PR DESCRIPTION
**Pull request by VamsiExcelra**
implemented the shared common component for table datagrid with sorting, filtering, row grouping, infinite scroll, columns ordering and options to use across application
implemented the api in route
updated the schema
integrated the api in frontend
handled the internal and external users.

### CODE REVIEW

File: apps/web/sharedComponents/table/dataGrid.tsx [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+                            <span className="flex items-center gap-[5px]">
+                                {data.status}
+                                <StatusMark status={data.status} />
+                            </span>
+                        )}
+                    />
+                );
+            }
+
+            if (field === "caco2" || field === "herg" || field === "clint" || field === "hepg2cytox" || field === "solubility") {
```
Comment (himanshubexcelra) [inline apps/web/sharedComponents/table/dataGrid.tsx:114]: Columns should come from component calling data grid. It's not fixed. 

Purpose of adding shared component is to use it everywhere in application. It's not onlt for molecule realted data grid.

I want to use it for users and organization table also

File: apps/web/utils/constants.ts [hunk 2]
```diff
@@ -237,3 +237,105 @@ export const features = [
   { name: "Multiple Org Access", value: "multiple_org_access", checked: false },
   { name: "Order Management", value: "order_management", checked: false },
 ];
+
+export const MoleculeOrders = [
```
Comment (himanshubexcelra) [inline apps/web/utils/constants.ts:241]: Fetch molecule order data from api & db.

File: apps/web/sharedComponents/table/dataGrid.tsx (hunk 1 above)
Comment (VamsiExcelra) [reply apps/web/sharedComponents/table/dataGrid.tsx]: changes done and updated

File: apps/web/utils/constants.ts (hunk 2 above)
Comment (VamsiExcelra) [reply apps/web/utils/constants.ts]: api implemented to fetch data from db

**Comment by VamsiExcelra** [comment]
1. implemented the shared common component for table datagrid with sorting, filtering, row grouping, infinite scroll, columns ordering and options to use across application
2. implemented the api in route
3. updated the schema
4. integrated the api in frontend
5. handled the internal and external users.

File: apps/web/app/molecule_order/page.tsx [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  }
+
+  const { userData } = sessionData;
+  const { myRoles, organizationId } = userData;
+
+  // Identify if the user is internal or external based on roles
+  const isInternalUser = myRoles.some((role: string) => 
+    role === 'researcher' || role === 'protocol_approver');
+  const isExternalUser = myRoles.some((role: string) => 
+    role === 'library_manager' || role === 'admin');
```
Comment (himanshubexcelra) [inline apps/web/app/molecule_order/page.tsx:24]: admin is not external user org_admin is external user.

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+  const breadcrumbs: BreadCrumbsObj[] = [
+    { label: "Home", svgPath: "/icons/home-icon.svg", svgWidth: 16, svgHeight: 16, href: "/" },
+    {
+      label: "Molecule Orders", svgPath: "/icons/molecule-order.svg",
+      svgWidth: 16, svgHeight: 16, href: "/projects"
+    },
+  ];
+
+  const handleBookMarkItem = (data: any) => console.log(data, 'data-data');
```
Comment (himanshubexcelra) [inline apps/web/components/MoleculeOrder/MoleculeOrder.tsx:63]: Remove all console log

File: apps/web/app/molecule_order/page.tsx [hunk 5]
```diff
@@ -0,0 +1,60 @@
+/*eslint max-len: ["error", { "code": 100 }]*/
+import Layout from '@/components/layout';
+import { getMoleculesOrder } from './service';
+import { getUserData } from '@/utils/auth';
+import { redirect } from 'next/navigation';
+import MoleculeOrderPage from '@/components/MoleculeOrder/MoleculeOrder';
+
+export default async function Page({ searchParams }: 
+  { searchParams: { projectId?: string; libraryId?: string } }) {
```
Comment (himanshubexcelra) [inline apps/web/app/molecule_order/page.tsx:9]: Define prop type properly. You can check in Header.tsx or LoginForm.tsx.

File: apps/web/app/molecule_order/page.tsx [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      data = await getMoleculesOrder({ organizationId });
+    } else if (isExternalUser && searchParams.projectId && searchParams.libraryId) {
+      // For external users, pass projectId and libraryId as parameters
+      data = await getMoleculesOrder({
+        projectId: searchParams.projectId,
+        libraryId: searchParams.libraryId,
+      });
+    } else {
+      console.warn("User role not supported or missing required parameters.");
+      return <MoleculeOrderPage initialData={[]} />;
```
Comment (himanshubexcelra) [inline apps/web/app/molecule_order/page.tsx:39]: There should be only one return statement. You should not decide layout conditionally. You should pass data conditionally to one returned layout.

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  smile: string;
+  status: string;
+}
+
+// Custom renderer function
+const customRenderForField = (data: MoleculeOrder, field: string) => {
+  let color: StatusCodeType = 'ready';
+  const value = data[field];
+
+  if (value !== undefined) {
```
Comment (himanshubexcelra) [inline apps/web/components/MoleculeOrder/MoleculeOrder.tsx:38]: This one i am not getting what you are trying to do here.

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+  const handleBookMarkItem = (data: any) => console.log(data, 'data-data');
+  const handleStructureEdit = (data: any) => console.log(data, 'edit');
+  const handleStructureZoom = (data: any) => console.log(data, 'edit');
+  const handleStructureDelete = (data: any) => console.log(data, 'edit');
+
+  // Column configuration
+  const moleculeColumns = [
+    {
+      dataField: 'bookmark',
+      width: 100,
```
Comment (himanshubexcelra) [inline apps/web/components/MoleculeOrder/MoleculeOrder.tsx:72]: In molecule order this column should not come i guess. I will confirm.

File: apps/web/sharedComponents/table/dataGrid.tsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
+import React, { useCallback, useEffect, useRef, useState } from 'react';
```
Comment (himanshubexcelra) [inline apps/web/sharedComponents/table/dataGrid.tsx:1]: eslint rule is missing here

File: apps/api/app/v1/molecule_order/route.tsx [hunk 10]
```diff
@@ -0,0 +1,65 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
[the remark this answers is not in the document]
Comment (VamsiExcelra) [reply apps/api/app/v1/molecule_order/route.tsx]: rule removed

File: apps/api/prisma/schema.prisma
[the remark this answers is not in the document]
Comment (VamsiExcelra) [reply apps/api/prisma/schema.prisma:1]: schema updated with main branch schema

File: apps/web/app/molecule_order/page.tsx [hunk 11]
```diff
@@ -0,0 +1,60 @@
+/*eslint max-len: ["error", { "code": 100 }]*/
+import Layout from '@/components/layout';
+import { getMoleculesOrder } from './service';
+import { getUserData } from '@/utils/auth';
+import { redirect } from 'next/navigation';
+import MoleculeOrderPage from '@/components/MoleculeOrder/MoleculeOrder';
+
+export default async function Page({ searchParams }:
```
[the remark this answers is not in the document]
Comment (VamsiExcelra) [reply apps/web/app/molecule_order/page.tsx:8]: page updated 'MoleculeOrder' name

File: apps/web/app/molecule_order/page.tsx [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+      // For internal users, pass organizationId as a parameter
+      data = await getMoleculesOrder({ organizationId });
+    } else if (isExternalUser && searchParams.projectId && searchParams.libraryId) {
+      // For external users, pass projectId and libraryId as parameters
+      data = await getMoleculesOrder({
+        projectId: searchParams.projectId,
+        libraryId: searchParams.libraryId,
+      });
+    } else {
+      console.warn("User role not supported or missing required parameters.");
```
[the remark this answers is not in the document]
Comment (VamsiExcelra) [reply apps/web/app/molecule_order/page.tsx:38]: moved to message.ts

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx (hunk 4 above)
Comment (VamsiExcelra) [reply apps/web/components/MoleculeOrder/MoleculeOrder.tsx:63]: all consoles removed

File: apps/web/app/molecule_order/page.tsx (hunk 5 above)
Comment (VamsiExcelra) [reply apps/web/app/molecule_order/page.tsx:9]: type added

File: apps/web/app/molecule_order/page.tsx (hunk 6 above)
Comment (VamsiExcelra) [reply apps/web/app/molecule_order/page.tsx:39]: changed to one return statement

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx (hunk 7 above)
Comment (VamsiExcelra) [reply apps/web/components/MoleculeOrder/MoleculeOrder.tsx:38]: Dynamic field access based on the `field` parameter

File: apps/web/app/molecule_order/page.tsx (hunk 3 above)
Comment (VamsiExcelra) [reply apps/web/app/molecule_order/page.tsx:24]: updated with user type

**Comment by himanshubexcelra** [comment]
image

see alignment and spacing is not according to figma screen.

**Comment by himanshubexcelra** [comment]
breadcrumb is not active.

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+  structure: string;
+  orderId: number;
+  orderName: string;
+  moleculeId: number;
+  molecularWeight: number;
+  organizationName: string;
+  molecular_weight: string;
+  smile: string;
+  status: string;
+  yield?: number;
```
Comment (himanshubexcelra) [inline apps/web/components/MoleculeOrder/MoleculeOrder.tsx:28]: these columns are not there in molecule_order.

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+  minWidth?: number;
+  customRender?: (data: T) => React.ReactNode;
+}
+
+// Custom renderer function
+const customRenderForField = (data: MoleculeOrder, field: keyof MoleculeOrder) => {
+  let color: StatusCodeAPIType = 'READY';
+  const value = data[field]; // Dynamic field access based on the `field` parameter
+
+  if (typeof value === 'number') {
```
Comment (himanshubexcelra) [inline apps/web/components/MoleculeOrder/MoleculeOrder.tsx:47]: Status design should be similar to figma and i think this code is not correct.

@ForumExcelra  @VamsiExcelra

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+  const breadcrumbs: BreadCrumbsObj[] = [
+    { label: 'Home', svgPath: '/icons/home-icon.svg', svgWidth: 16, svgHeight: 16, href: '/' },
+    {
+      label: 'Molecule Orders', svgPath: '/icons/molecule-order.svg',
+      svgWidth: 16, svgHeight: 16, href: '/projects'
+    },
+  ];
+
+  const handleBookMarkItem = () => { };
```
Comment (himanshubexcelra) [inline apps/web/components/MoleculeOrder/MoleculeOrder.tsx:72]: These all constants not needed. Remove it.

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+    },
+    { dataField: 'orderId', title: 'Order' },
+    { dataField: 'moleculeId', title: 'Molecule ID' },
+    { dataField: 'molecular_weight', title: 'Molecular Weight' },
+    {
+      dataField: 'status',
+      title: 'Status',
+      width: 170,
+      customRender: (data: MoleculeOrder) => {
+        const colorKey = data.status.toUpperCase() as keyof typeof StatusCodeBg;
```
Comment (himanshubexcelra) [inline apps/web/components/MoleculeOrder/MoleculeOrder.tsx:114]: Complete status code looks wrong.

@ForumExcelra  @VamsiExcelra

**Comment by himanshubexcelra** [comment]
Send to Synthesis button is not yet added.

**Comment by himanshubexcelra** [comment]
As  lib manage i created orders. when i come to molecule orders page i am not able to see orders.
image

There are orders under organization 2 which is fauxbio and for that alos it's empty.

image

**Comment by himanshubexcelra** [comment]
image
Why external type is added in lowercase.



As per the number of comments increasing . I can say unit testing is not done properly on this PR. Just by assumption you are adding things. 4 story points has reached to 9 story points.

**Comment by himanshubexcelra** [comment]
Why here type props is not defined like i explained many times.
image
Whatever i am explaining it's repeating multiple times and still it's not done.

**Comment by himanshubexcelra** [comment]
In shared data grid component why Send for synthesis popup is added.

image

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx (hunk 13 above)
Comment (VamsiExcelra) [reply apps/web/components/MoleculeOrder/MoleculeOrder.tsx:37]: all the mentioned columns are there in figma design and for 'molecular_weight' i am displaying value too which is coming from 'molecules'

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx (hunk 14 above)
Comment (VamsiExcelra) [reply apps/web/components/MoleculeOrder/MoleculeOrder.tsx:56]: color codes are i have checked

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx (hunk 15 above)
Comment (VamsiExcelra) [reply apps/web/components/MoleculeOrder/MoleculeOrder.tsx]: removed

File: apps/web/components/MoleculeOrder/MoleculeOrder.tsx (hunk 16 above)
Comment (VamsiExcelra) [reply apps/web/components/MoleculeOrder/MoleculeOrder.tsx:169]: color codes are i have checked

**Comment by VamsiExcelra** [comment]
> Send to Synthesis button is not yet added.

send to synthesis button added

**Pull request merged**